File: ruby/543.diameter-of-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def diameter_of_binary_tree(root)
  @max = 0
  diameter(root)

  @max
end


def diameter(node)
  return -1 if node.nil?

  right = diameter(node.right) + 1
  left = diameter(node.left) + 1
  local_max = right + left

  # 後置if の方がmaxよりも処理が早い
  @max = local_max if local_max > @max

  # 三項演算子 の方がmaxよりも処理が早い
  left > right ? left : right
end

File: rust/src/543.diameter-of-binary-tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::cmp;
use std::rc::Rc;

impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut max = 0;
        Self::diameter(&root, &mut max);
        max
    }

    fn diameter(node: &Option<Rc<RefCell<TreeNode>>>, max: &mut i32) -> i32 {
        if let Some(node) = node {
            let node = node.borrow();
            let left = Self::diameter(&node.left, max);
            let right = Self::diameter(&node.right, max);

            *max = cmp::max(*max, left + right);
            cmp::max(left, right) + 1
        } else {
            0
        }
    }
}
